File: vnext/Microsoft.ReactNative.Cxx/JSValueTreeReader.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "JSValueTreeReader.h"

namespace winrt::Microsoft::ReactNative {

//===========================================================================
// JSValueTreeReader implementation
//===========================================================================

JSValueTreeReader::StackEntry::StackEntry(const JSValue &value, const JSValueObject::const_iterator &property) noexcept
    : Value{value}, Property{property} {}

JSValueTreeReader::StackEntry::StackEntry(const JSValue &value, const JSValueArray::const_iterator &item) noexcept
    : Value{value}, Item{item} {}

JSValueTreeReader::JSValueTreeReader(const JSValue &value) noexcept : m_root{value}, m_current{&value} {}

JSValueTreeReader::JSValueTreeReader(JSValue &&value) noexcept
    : m_ownedValue{std::move(value)}, m_root{m_ownedValue}, m_current{&m_ownedValue} {}

JSValueType JSValueTreeReader::ValueType() noexcept {
  return m_current->Type();
}

bool JSValueTreeReader::GetNextObjectProperty(hstring &propertyName) noexcept {
  if (!m_isInContainer) {
    if (m_current->Type() == JSValueType::Object) {
      const auto &properties = m_current->Object();
      const auto &property = properties.begin();
      if (property != properties.end()) {
        m_stack.emplace_back(*m_current, property);
        SetCurrentValue(property->second);
        propertyName = to_hstring(property->first);
        return true;
      } else {
        m_isInContainer = !m_stack.empty();
      }
    }
  } else if (!m_stack.empty()) {
    auto &entry = m_stack.back();
    if (entry.Value.Type() == JSValueType::Object) {
      auto &property = entry.Property;
      if (++property != entry.Value.Object().end()) {
        SetCurrentValue(property->second);
        propertyName = to_hstring(property->first);
        return true;
      } else {
        m_current = &entry.Value;
        m_stack.pop_back();
        m_isInContainer = !m_stack.empty();
      }
    }
  }

  propertyName = to_hstring(L"");
  return false;
}

bool JSValueTreeReader::GetNextArrayItem() noexcept {
  if (!m_isInContainer) {
    if (m_current->Type() == JSValueType::Array) {
      const auto &item = m_current->Array().begin();
      if (item != m_current->Array().end()) {
        m_stack.emplace_back(*m_current, item);
        SetCurrentValue(*item);
        return true;
      } else {
        m_isInContainer = !m_stack.empty();
      }
    }
  } else if (!m_stack.empty()) {
    auto &entry = m_stack.back();
    if (entry.Value.Type() == JSValueType::Array) {
      if (++entry.Item != entry.Value.Array().end()) {
        SetCurrentValue(*entry.Item);
        return true;
      } else {
        m_current = &entry.Value;
        m_stack.pop_back();
        m_isInContainer = !m_stack.empty();
      }
    }
  }

  return false;
}

void JSValueTreeReader::SetCurrentValue(const JSValue &value) noexcept {
  m_current = &value;
  switch (value.Type()) {
    case JSValueType::Object:
    case JSValueType::Array:
      m_isInContainer = false;
      break;
    default:
      m_isInContainer = true;
      break;
  }
}

hstring JSValueTreeReader::GetString() noexcept {
  return to_hstring((ValueType() == JSValueType::String) ? m_current->String() : "");
}

bool JSValueTreeReader::GetBoolean() noexcept {
  return (ValueType() == JSValueType::Boolean) ? m_current->Boolean() : false;
}

int64_t JSValueTreeReader::GetInt64() noexcept {
  return (ValueType() == JSValueType::Int64) ? m_current->Int64() : 0;
}

double JSValueTreeReader::GetDouble() noexcept {
  return (ValueType() == JSValueType::Double) ? m_current->Double() : 0;
}

IJSValueReader MakeJSValueTreeReader(const JSValue &root) noexcept {
  return make<JSValueTreeReader>(root);
}

IJSValueReader MakeJSValueTreeReader(JSValue &&root) noexcept {
  return make<JSValueTreeReader>(std::move(root));
}

} // namespace winrt::Microsoft::ReactNative

File: packages/react-native-windows/Microsoft.ReactNative.Managed/JSValueTreeReader.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Collections.Generic;

namespace Microsoft.ReactNative.Managed
{
  public class JSValueTreeReader : IJSValueReader, IJSValueTreeReader
  {
    private bool m_isIterating = false;
    private readonly Stack<StackEntry> m_stack = new Stack<StackEntry>();

    public JSValueTreeReader(JSValue root)
    {
      Root = root;
      Current = root;
    }

    public JSValueType ValueType => Current.Type;

    public JSValue Current { get; private set; }

    public JSValue Root { get; private set; }

    public bool GetNextObjectProperty(out string propertyName)
    {
      if (!m_isIterating)
      {
        if (Current.TryGetObject(out var currentObject))
        {
          var properties = currentObject.GetEnumerator();
          if (properties.MoveNext())
          {
            m_stack.Push(StackEntry.ObjectProperty(Current, properties));
            SetCurrentValue(properties.Current.Value);
            propertyName = properties.Current.Key;
            return true;
          }
          else
          {
            m_isIterating = m_stack.Count != 0;
          }
        }
      }
      else if (m_stack.Count != 0)
      {
        StackEntry entry = m_stack.Peek();
        if (entry.Value.Type == JSValueType.Object)
        {
          if (entry.Property.MoveNext())
          {
            SetCurrentValue(entry.Property.Current.Value);
            propertyName = entry.Property.Current.Key;
            return true;
          }
          else
          {
            Current = entry.Value;
            m_stack.Pop();
            m_isIterating = m_stack.Count != 0;
          }
        }
      }

      propertyName = "";
      return false;
    }

    public bool GetNextArrayItem()
    {
      if (!m_isIterating)
      {
        if (Current.TryGetArray(out var currentArray))
        {
          var items = currentArray.GetEnumerator();
          if (items.MoveNext())
          {
            m_stack.Push(StackEntry.ArrayItem(Current, items));
            SetCurrentValue(items.Current);
            return true;
          }
        }
        else
        {
          m_isIterating = m_stack.Count != 0;
        }

      }
      else if (m_stack.Count != 0)
      {
        var entry = m_stack.Peek();
        if (entry.Value.Type == JSValueType.Array)
        {
          if (entry.Item.MoveNext())
          {
            SetCurrentValue(entry.Item.Current);
            return true;
          }
          else
          {
            Current = entry.Value;
            m_stack.Pop();
            m_isIterating = m_stack.Count != 0;
          }
        }
      }

      return false;
    }

    private void SetCurrentValue(JSValue value)
    {
      Current = value;
      switch (value.Type)
      {
        case JSValueType.Object:
        case JSValueType.Array:
          m_isIterating = false;
          break;
        default:
          m_isIterating = true;
          break;
      }
    }

    public string GetString()
    {
      return Current.TryGetString(out var stringValue) ? stringValue : "";
    }

    public bool GetBoolean()
    {
      return Current.TryGetBoolean(out var booleanValue) ? booleanValue : false;
    }

    public long GetInt64()
    {
      return Current.TryGetInt64(out var int64Value) ? int64Value : 0;
    }

    public double GetDouble()
    {
      return Current.TryGetDouble(out var doubleValue) ? doubleValue : 0;
    }

    private struct StackEntry
    {
      public JSValue Value;
      public IEnumerator<JSValue> Item;
      public IEnumerator<KeyValuePair<string, JSValue>> Property;

      public static StackEntry ObjectProperty(JSValue value, IEnumerator<KeyValuePair<string, JSValue>> property)
      {
        return new StackEntry
        {
          Value = value,
          Item = null,
          Property = property
        };
      }

      public static StackEntry ArrayItem(JSValue value, IEnumerator<JSValue> item)
      {
        return new StackEntry
        {
          Value = value,
          Item = item,
          Property = null
        };
      }
    }
  }
}
